### PR TITLE
Use simpler method to detect binaries

### PR DESCRIPTION
**Pull request by danielnachun**
We discussed here: https://github.com/Homebrew/brew/pull/12940/files#r818284011 that it is much simpler to use `grep` to check if an individual file contains null bytes than to iterate over all files in the keg.  We are only interested in checking files which we already know contain prefixes, which we can pass in as an argument.  This method could also be useful in the future if we need to determine if files are binaries in other situations.

I also reverted `each_unique_file_matching` back to its original state.  If we think it's better to still have a generic `each_unique_file` method that is called by `each_unique_file_matching`, that's totally fine with me.  At the moment only `each_unique_file_matching` would need to call `each_unique_file`, but it's possible some future methods could also use `each_unique_file` as well.

### CODE REVIEW

File: Library/Homebrew/extend/os/mac/keg_relocate.rb [hunk 1]
```diff
@@ -186,7 +186,7 @@ def recursive_fgrep_args
 
   def egrep_args
     grep_bin = "egrep"
-    grep_args = recursive_fgrep_args
+    grep_args = "-l"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/mac/keg_relocate.rb:189]: Are there long flags you can use here so it's a little more readable?

File: Library/Homebrew/keg_relocate.rb [hunk 2]
```diff
@@ -177,41 +177,34 @@ def recursive_fgrep_args
 
   def egrep_args
     grep_bin = "grep"
-    grep_args = recursive_fgrep_args
-    grep_args += "Pa"
+    grep_args = "-Pla"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/keg_relocate.rb:180]: Are there long flags you can use here so it's a little more readable?

File: Library/Homebrew/keg_relocate.rb [hunk 3]
```diff
@@ -177,41 +177,34 @@ def recursive_fgrep_args
...
+  def binary_file?(file)
     grep_bin, grep_args = egrep_args
 
     # We need to pass NULL_BYTE_STRING, the literal string "\x00", to grep
     # rather than NULL_BYTE, a literal null byte, because grep will internally
     # convert the literal string "\x00" to a null byte.
-    Utils.popen_read(grep_bin, grep_args, NULL_BYTE_STRING, to_s) do |io|
-      each_unique_file(io, block)
-    end
+    Utils.popen_read(grep_bin, grep_args, NULL_BYTE_STRING, file).size.positive?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/keg_relocate.rb:207]: Will this stop when it reads the first null byte?

**Review by MikeMcQuaid — APPROVED** [review]
A few thoughts but fine as-is, thanks!

File: Library/Homebrew/keg_relocate.rb (hunk 3 above)
Comment (danielnachun) [reply Library/Homebrew/keg_relocate.rb:207]: Yes, the "-l" argument we use on both OSes is short for "--files-with-matches", which according the manuals for GNU and BSD grep will stop scanning after the first match.  Using the long names of the arguments as you suggested above will help to clarify this.

File: Library/Homebrew/keg_relocate.rb (hunk 2 above)
Comment (danielnachun) [reply Library/Homebrew/keg_relocate.rb]: I've switched the flags to now use the long versions.

File: Library/Homebrew/keg_relocate.rb [hunk 4]
```diff
@@ -177,41 +177,36 @@ def recursive_fgrep_args
 
   def egrep_args
     grep_bin = "grep"
-    grep_args = recursive_fgrep_args
-    grep_args += "Pa"
+    grep_args = ["--files-with-matches",
+                 "--perl-regexp",
+                 "--binary-files=text"]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/keg_relocate.rb:182]:
```suggestion
    grep_args = [
      "--files-with-matches",
      "--perl-regexp",
      "--binary-files=text",
    ]
```

just if we decide to add/remove this later to make the diff nicer.

File: Library/Homebrew/keg_relocate.rb [hunk 5]
```diff
@@ -177,41 +177,36 @@ def recursive_fgrep_args
...
+  def binary_file?(file)
     grep_bin, grep_args = egrep_args
 
     # We need to pass NULL_BYTE_STRING, the literal string "\x00", to grep
     # rather than NULL_BYTE, a literal null byte, because grep will internally
     # convert the literal string "\x00" to a null byte.
-    Utils.popen_read(grep_bin, grep_args, NULL_BYTE_STRING, to_s) do |io|
-      each_unique_file(io, block)
-    end
+    Utils.popen_read(grep_bin, *grep_args, NULL_BYTE_STRING, file).size.positive?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/keg_relocate.rb:209]:
```suggestion
    Utils.popen_read(grep_bin, *grep_args, NULL_BYTE_STRING, file).present?
```

**Review by MikeMcQuaid — APPROVED** [review]
Two optional style suggestions, feel free to ship as-is or after those have been applied. Thanks @danielnachun!

**Comment by danielnachun** [comment]
I made the style changes and am merging now.

**Pull request merged**